launcher.py

```diff
@@ -12,94 +12,149 @@
-            cmd = "ping -c 1 8.8.8.8"
-        logging.debug("Checking internet connectivity... ")
-        run_command(cmd)
-        logging.debug("Internet connectivity OK")
-        return True
+    @staticmethod
+    def run_command(command: str, cwd: Optional[Path] = None):
+        """
+        Executes a shell command.
 
-    except Exception as e:
-        logging.error("No internet connection detected.")
-        logging.debug(f"Error: {e}")
-        return False
+        Args:
+            command (str): The command to run.
+            cwd (str, optional): The working directory. Defaults to None.
+
+        Raises:
+            subprocess.CalledProcessError: If any command fails.
+        """
+        logging.debug(f"Executing command: {command}")
+        process = subprocess.Popen(
+            command.split(),
+            cwd=cwd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            text=True,
+        )
+        stdout, stderr = process.communicate()  # wait for process to terminate
+        if stdout:
+            logging.debug(f"Command output: {stdout}")
+        if stderr:
+            logging.error(f"Command error: {stderr}")
+
+        if process.returncode != 0:
+            raise subprocess.CalledProcessError(process.returncode, command)
+
+    @staticmethod
+    def check_internet_connectivity(system) -> bool:
+        try:
+            cmd = "ping -n 1 8.8.8.8" if system == "Windows" else "ping -c 1 8.8.8.8"
+            Utils.run_command(cmd)
+            logging.debug("Internet connectivity OK")
+            return True
+        except Exception as e:
+            logging.error("No internet connection detected.")
+            logging.debug(f"Error: {e}")
+            return False
+
+    @staticmethod
+    def ensure_virtualenv_exists():
+        if Utils.VENV_DIR_PATH.exists():
+            logging.debug("Virtual environment already exists.")
+            return
+        logging.info(f"Generating virtual environment at: {Utils.VENV_DIR_PATH}")
+        Utils.run_command(f"{sys.executable} -m venv {str(Utils.VENV_DIR_PATH)} --system-site-packages")
+        logging.info("Successfully generated Virtual environment.")
+
+    @staticmethod
+    def get_virtualenv_folder() -> Path:
+        """
+        get the path of the virtual environment binaries folder
+        """
+        Utils.ensure_virtualenv_exists()
+        for folder in ["bin", "Scripts"]:
+            path = Utils.VENV_DIR_PATH / folder
+            logging.debug(f"Searching VENV Path: {path}")
+            if path.exists():
+                return path
+
+        logging.error("Could not locate virtual environment folders.")
+        sys.exit(1)
 
 
 def install_msys2():
+    """
+    Installs MSYS2 and Update Path enviornment variable for windows platform
+    """
     try:
         installer_url = (
             "https://github.com/msys2/msys2-installer/releases/download/nightly-x86_64/msys2-base-x86_64-latest.sfx.exe"
```

Comment:
I've been meaning to ask this: any reason why you're using the nightly version? Not familiar with it myself, but don't they have a stable release? At least good to see the switch to `latest` instead of a specific version hash...

Reply:
changed this to stable.